### PR TITLE
Refactor check valve with spline interpolation

### PR DESCRIPTION
**Pull request by AntoineGautier**
This addresses #1937.

### CODE REVIEW

**Comment by AntoineGautier** [comment]
@mwetter This is ready. Can you suggest a name for an external review?

**Comment by AntoineGautier** [comment]
@mwetter This is ready. Can you suggest a name for an external review?

File: IBPSA/Fluid/FixedResistances/CheckValve.mo [hunk 1]
```diff
@@ -27,48 +27,75 @@ model CheckValve "Check valve that avoids flow reversal"
...
+      x=dp,
+      x1=0,
+      x2=dpValve_closing,
+      y1=m_flow_min,
+      y2=m_flow_max,
+      y1d=IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_dp_der(
+        dp=0,
+        k=k_min,
+        m_flow_turbulent=m_flow_turbulent,
+        dp_der=1),
```
Comment (mwetter) [inline IBPSA/Fluid/FixedResistances/CheckValve.mo:75]: Can you explain why `dp_der=1` and `dp_der2=0`. I don't understand what you implemented here.
(A side note is that these are parameter expressions so for efficiency that could be computed as a parameter.)

**Review by mwetter — COMMENTED** [review]
@AntoineGautier : I am not sure we approached this the right way, in part because I don't understand your implementation (see inline comment).

Wouldn't it be simpler if we were to compute $dp$ from $\dot m$? This would make it simpler because both components have the same $\dot m$, so we can simply add their resistances. For the fixed resistance the evaluation is easy (with the flow function). For the actual valve, we can make a linear behavior for $\dot m \le 0$ and quadratic for $\dot m \ge \dot m_0/2$. In between we use a polynominal that fits the first and second derivatives at these points. The coefficients of the polynomial can be computed as parameter expressions.

File: IBPSA/Fluid/BaseClasses/FlowModels/basicFlowFunction_dp_der.mo [hunk 2]
```diff
@@ -31,7 +31,7 @@ Documentation(info="<html>
 Function that implements the first order derivative of
 <a href=\"modelica://IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_dp\">
 IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_dp</a>
-with respect to the mass flow rate.
+with respect to time.
```
[the remark this answers is not in the document]
Comment (AntoineGautier) [reply IBPSA/Fluid/BaseClasses/FlowModels/basicFlowFunction_dp_der.mo:34]: Considering that
- the algorithm of the function `basicFlowFunction_dp_der` can be written as: 
$`m\_flow\_der = \frac{\partial basicFlowFunction\_dp}{\partial dp} * \frac{\partial dp}{\partial t}`$ (where $`\frac{\partial dp}{\partial t} = dp\_der`$)
- its output is declared with kg/s2 unit: `m_flow_der(unit="kg/s2")`
- the function is validated in `Fluid.BaseClasses.FlowModels.Validation.BasicFlowFunction_dp_DerivativeCheck` by comparing the time integral of the derivative `der(m_flow_comp)` with the original function `basicFlowFunction_dp`

I conclude that `basicFlowFunction_dp_der` is the derivative of `basicFlowFunction_dp` with respect to *time* (not `dp`).

Then, considering 
$`\frac{\partial m\_flow}{\partial t} = \frac{\partial m\_flow}{\partial dp} * \frac{\partial dp}{\partial t}`$
I compute $`\frac{\partial m\_flow}{\partial dp}`$ by calling `basicFlowFunction_dp_der(..., dp_der=1)`.

File: IBPSA/Fluid/FixedResistances/CheckValve.mo (hunk 1 above)
Comment (AntoineGautier) [reply IBPSA/Fluid/FixedResistances/CheckValve.mo:75]: See above.
For the second derivative, considering
$`\frac{\partial^2 m\_flow}{\partial t^2} = \frac{\partial^2 m\_flow}{\partial dp^2} * \left(\frac{\partial dp}{\partial t}\right)^2 + \frac{\partial m\_flow}{\partial dp} * \frac{\partial^2 dp}{\partial t^2}`$
I compute $`\frac{\partial^2 m\_flow}{\partial dp^2}`$ by calling `basicFlowFunction_dp_der2(..., dp_der=1, dp_der2=0)`.

**Comment by AntoineGautier** [comment]
@mwetter With [11097a8](https://github.com/ibpsa/modelica-ibpsa/pull/1938/commits/11097a81bf2b34fb2c77ec733fd70d8ae9dd70be) the function calls are now included in the `if else` clauses, `min` and `max` are removed and parameters are introduced when possible.
I've also updated the documentation for the derivatives of the basic flow functions as discussed. Note the important information that these derivatives are formulated under the assumption of a constant flow coefficient. This is consistent with the annotation `derivative(zeroDerivative=k)` in `basicFlowFunction_dp`. This also means that if a model features a varying flow coefficient (e.g., any actuator model), these derivative functions will not be used by the tools, which will rely on automatic differentiation instead.

**Review by mwetter — APPROVED** [review]
@AntoineGautier : Thanks for the changes. This addresses my concerns that we discussed on Friday.

**Pull request merged**